src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1211,14 +1213,21 @@ void MergeTreeData::clearPartsFromFilesystem(const DataPartsVector & parts_to_re
     {
         /// Parallel parts removal.
 
-        size_t num_threads = std::min(size_t(settings->max_part_removal_threads), parts_to_remove.size());
+        size_t num_threads = std::min<size_t>(settings->max_part_removal_threads, parts_to_remove.size());
         ThreadPool pool(num_threads);
 
         /// NOTE: Under heavy system load you may get "Cannot schedule a task" from ThreadPool.
         for (const DataPartPtr & part : parts_to_remove)
         {
-            pool.scheduleOrThrowOnError([&]
+            pool.scheduleOrThrowOnError([&, thread_group = CurrentThread::getGroup()]
             {
+                SCOPE_EXIT_SAFE(
```

Comment:
BTW we don't have this in all other cases when we attaching thread group. Is it really needed? Is it needed in other places?

Reply:
If you are talking about SCOPE_EXIT_SAFE (safe variant of SCOPE_EXIT) then yes, since detachQuery looks pretty complex and it may throw.

If you are talking about "why do we need detachQuery" then the answer is also yes, since it responsible for proper state  for the thread and finalizes things (query profiler, performance counters, opentelementry and so on)

Reply:
I will take a look at a more generic approach for threads accounting later.

Reply:
No need to, there is #22972 already

Reply:
@azat, #22972 will not be merged, there is huge linker issues, and maybe some sort of interface split will be required to make it easy. For now  SCOPE_EXIT_SAFE is fine.
CC: @alexey-milovidov